File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/Text.java
// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data;

import com.daml.ledger.api.v1.ValueOuterClass;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.Objects;

public class Text extends Value {

    private final String value;

    public Text(@NonNull String value) {
        this.value = value;
    }

    @NonNull
    public String getValue() {
        return value;
    }

    @Override
    public ValueOuterClass.Value toProto() {
        return ValueOuterClass.Value.newBuilder().setText(this.value).build();
    }

    @Override
    public String toString() {
        return "Text{" +
                "value='" + value + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Text text = (Text) o;
        return Objects.equals(value, text.value);
    }

    @Override
    public int hashCode() {

        return Objects.hash(value);
    }
}

File: sdk/canton/community/daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Text.scala
// Copyright (c) 2026 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.digitalasset.daml.lf.data

sealed abstract class TextModule {

  type Text <: String

  def fromString(s: String): Either[String, Text]

  final def assertFromString(s: String) = assertRight(fromString(s))

}

private[data] object TextModuleImpl extends TextModule {
  override type Text = String
  override def fromString(s: String): Either[String, Text] = {
    val length = s.length

    @scala.annotation.tailrec
    def loop(i: Int): Either[String, Text] =
      if (i >= length) Right(new Text(s))
      else {
        val ch = s.charAt(i)

        // Check for null character
        if (ch == '\u0000')
          Left("Text contains null character")
        // Check for valid surrogate pairs
        else if (Character.isHighSurrogate(ch)) {
          if (i + 1 >= length || !Character.isLowSurrogate(s.charAt(i + 1))) {
            Left("Text contains unpaired high surrogate")
          } else {
            loop(i + 2) // Skip the low surrogate
          }
        } else if (Character.isLowSurrogate(ch)) {
          Left("Text contains unpaired low surrogate")
        } else {
          loop(i + 1)
        }
      }

    loop(0)
  }
}
